cmapPy/math/fast_cov.py

```diff
@@ -0,0 +1,87 @@
+import logging
+import cmapPy.pandasGEXpress.setup_GCToo_logger as setup_logger
+import numpy
+
+
+logger = logging.getLogger(setup_logger.LOGGER_NAME)
+
+
+def fast_cov(x, y=None, axis=0, do_print_trace=False):
+    """calculate the covariance matrix for x, or optionally, the covariance matrix between x and y.
+    axis parameter defines whether to treat the rows or columns as vectors that the correlation is calculated
+    between
+
+    Args:
+        x (numpy array-like) NxM in shape
+        y (numpy array-like) OxP in shape
+        axis (integer) indicates which axis of the arrays is the vectors.  axis=0 indicates calculate
+            the covariance between the rows, axis=1 indicates calculate the covariance between the columns
+        do_print_trace (bool) whether or not to use logger.debug to print out intermediate mathematical steps
+            (WARNING: will print out all intermediate matrices, potentially lots of output!)
+
+        returns (numpy array-like) QxR array of the covariance values
+            for defaults (y=None, axis=0), shape is NxN
+            if y=None and axis=1, shape is MxM
+            if y is provied, axis=0, shape is NxO (and P==M in other words number of columns in x & y are the same)
+            if y is provided, axis=1, shape is MxP (and N==O in other words number of rows in x & y are the same)
+    """
+    validate_axis(axis)
+    validate_x_y(x, y, axis)
+
+    if y is None:
+        y = x
+
+    mean_x = numpy.mean(x.T, axis=axis)
```

Comment:
I think the problem is that the direction of operation is ambiguous when doing a correlation operation.  It is not ambiguous for operations like `sum` `mean` `median` etc.  Interestingly, pandas and numpy don't use the axis keyword for correlation and covariance (numpy uses `rowvar`; pandas just does correlation between the columns) - I suspect b/c of this ambiguity.
Numpy definition of rowvar is
> If rowvar is True (default), then each row represents a variable, with observations in the columns. Otherwise, the relationship is transposed: each column represents a variable, while the rows contain observations.
I believe that's using standard statistics language, where for us variable == experiment, observation == gene/peptide/histone mark/image feature etc.

So I think I should either change to using rowvar, or simplify even further and just only do it between the columns (in this latter case the user can just transpose as needed).  What do you think?

Reply:
Only supporting column correlation is not a bad idea. To use these functions, you have to be in Python anyway, so transposing is trivial.

If you're on command line though, say trying to correlate the columns of a GCT with a wrapper like `corr_gctoo.py`, a boolean argument like `--do_rows` could be nice, since transposing is not as easy.

Reply:
cool, if @oena agrees I'll make it only support columns.  For the command line it shouldn't be a problem to have that option, and then after it reads the GCT's in, it can do the transpose, amirght?

Reply:
Seems reasonable to me!